### PR TITLE
Deeply rework WeakIdKeyDictionary

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90825
* #90628

In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: test/test_weak.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+        dict = WeakIdKeyDictionary({o: 364})
+        self.assertEqual(dict[o], 364)
+
+    def test_make_weak_keyed_dict_from_weak_keyed_dict(self):
+        o = torch.randn(3)
+        dict = WeakIdKeyDictionary({o: 364})
+        dict2 = WeakIdKeyDictionary(dict)
+        self.assertEqual(dict[o], 364)
+
+    def make_weak_keyed_dict(self):
```
Comment (albanD) [inline test/test_weak.py:35]: This is not used?

File: test/test_weak.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+from torch.utils.weak import WeakIdKeyDictionary
+
+
+def C():
+    return torch.randn(1)
+
+
+# These tests are ported from cpython/Lib/test/test_weakref.py,
+# but adapted to use tensor rather than object
+class WeakTest(TestCase):
```
Comment (albanD) [inline test/test_weak.py:21]: There is no test in there that ensures that we only hold a weak reference and not a strong one?
Also any test to ensure we don't introduce ref cycles (which would delay deletion in practice)?

**Review by albanD — APPROVED** [review]
This solves https://github.com/pytorch/pytorch/issues/81597 ?

Sounds good!

File: torch/utils/weak.py [hunk 3]
```diff
@@ -1,79 +1,263 @@
...
+        d = self.data
+        self._pending_removals = [k for k in self._pending_removals if k in d]
+        self._dirty_len = False
+
+    def __delitem__(self, key):
+        self._dirty_len = True
+        del self.data[WeakIdRef(key)]  # CHANGED
+
+    def __getitem__(self, key):
+        return self.data[WeakIdRef(key)]  # CHANGED
```
Comment (eellison) [inline torch/utils/weak.py:131]: we need to do `_fix_weakref` here. You're calling it elsewhere but if we're going to expose this as a utility we shouldn't require users to call it

File: torch/utils/weak.py [hunk 4]
```diff
@@ -1,79 +1,263 @@
...
+        self._dirty_len = True
+        while True:
+            key, value = self.data.popitem()
+            o = key()
+            if o is not None:
+                return o, value
+
+    def pop(self, key, *args):
+        self._dirty_len = True
+        return self.data.pop(WeakIdRef(key), *args)  # CHANGED
```
Comment (eellison) [inline torch/utils/weak.py:221]: + 1, _fix_weakref, right ?

**Review by eellison — COMMENTED** [review]
this is not the same as the prior implementation, because the prior implementation was weak on both key and value

File: test/test_weak.py (hunk 2 above)
Comment (ezyang) [reply test/test_weak.py:21]: Weak reference is indirectly tested, as we do test that entries are promptly removed from dictionary when they die (this is how I found _fix_weakref leak).

There were some ref cycle tests in cpython that seemed complicated, but I could try to intro them

**Comment by ezyang** [comment]
I don't think so, pretty sure old impl was strong on value

File: torch/utils/weak.py (hunk 3 above)
Comment (ezyang) [reply torch/utils/weak.py:131]: I disagree. I never take out a strong reference to the key here, so I don't need to fix weakref

Comment (eellison) [reply torch/utils/weak.py:131]: You're right. Was thinking that data was a WeakValueDictionary

File: test/test_weak.py (hunk 2 above)
Comment (ezyang) [reply test/test_weak.py:21]: leaving this as follow up, I think prompt deletion also covers the ref cycle case

File: torch/utils/weak.py [hunk 5]
```diff
@@ -1,79 +1,263 @@
...
+        # cache the id of the key as we know this is definitely the hash
+        # method
+        self._id = id(key)
+        super().__init__(key, callback)
+
+    def __call__(self):
+        r = super().__call__()
+        # Special logic for Tensor PyObject resurrection
+        if hasattr(r, '_fix_weakref'):
+            r._fix_weakref()  # type: ignore[union-attr]
```
Comment (danthe3rd) [inline torch/utils/weak.py:53]: This class seems to still keep a strong reference to the object. Repro:
```python
import torch
import gc
import weakref
from _torch_weakkeyid import WeakTensorKeyDictionary, WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>

```

Commenting out "r._fix_weakref()" here fixes it

Comment (ezyang) [reply torch/utils/weak.py:53]: Do you have the first PR in this stack? There was a memory leak that I fixed

Comment (ezyang) [reply torch/utils/weak.py:53]: yeah on master, fixing your example, it works

```
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ cat foo.py
import torch
import gc
import weakref
from torch.utils.weak import WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ python foo.py
refX:  <class 'NoneType'>
refY:  <class 'NoneType'>
```

Comment (danthe3rd) [reply torch/utils/weak.py:53]: Oh right, thanks a lot! I just imported the `weak.py` file (as I wanted it to work with older versions of pytorch as well). I'll look into the other PR. Thanks!

Comment (ezyang) [reply torch/utils/weak.py:53]: For old versions just get rid of the fix weakref call, it's super busted lol. I suppose if you are willing to write a little C extension code, manually decreffing the Tensor after fixing weakref would work too

**Pull request closed without merging**